clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -290,17 +315,47 @@ public void poll(long timeout) throws IOException {
 
         clear();
 
-        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty())
+        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty() || !keysWithBytesBuffered.isEmpty())
             timeout = 0;
 
+        boolean poolOutOfMemory = memoryPool.isOutOfMemory();
+        if (poolOutOfMemory && !outOfMemory) {
+            //dont bother selecting for reads
+            outOfMemory = true;
+            log.trace("Broker is low on memory - muting incoming sockets");
+            for (KafkaChannel channel : channels.values()) {
+                if (channel.isInMutableState())
+                    channel.mute();
+            }
```

Comment:
It's just more proactive, right? If you're out of memory, you're not going to be able to do anything (beyond the handshake) on any channel anyway. I think the value is that instead of going through a more polling unnecessarily only to end up muting all the channels, you can just do so immediately.